jaxopt/_src/zoom_linesearch.py

```diff
@@ -272,21 +276,24 @@ def zoom_linesearch(f, xk, pk, old_fval=None, old_old_fval=None, gfk=None, c1=1e
   Returns: LineSearchResults
   """
   #xk, pk = _promote_dtypes_inexact(xk, pk)
-  xk = jnp.asarray(xk)
-  pk = jnp.asarray(pk)
+  #xk = jnp.asarray(xk)
+  #pk = jnp.asarray(pk)
 
   def restricted_func_and_grad(t):
-    t = jnp.array(t, dtype=pk.dtype)
-    # FIXME: directly acept a value_and_grad function to avoid recompilations.
-    phi, g = jax.value_and_grad(f)(xk + t * pk)
-    dphi = jnp.real(_dot(g, pk))
+    #t = jnp.array(t, dtype=pk.dtype)
+
+    xkp1 = tree_add_scalar_mul(xk, t, pk)
+
+    # FIXME: directly accept a value_and_grad function to avoid recompilations.
+    phi, g = jax.value_and_grad(f)(xkp1)
+    dphi = tree_real(tree_vdot(g, pk))
```

Comment:
How often do we have this pattern? I am not sure how expensive is the additional tree map, maybe it is better to a single function that we map to the input?
```python
real_vdot = lambda x, y: jnp.real(_vdot_safe(x, y))
tree_real_vdot = lambda x, y: tree_map(real_vdot, tree)
```

Reply:
Good question. I wonder if JAX's `jit` is able to optimize this automatically. If not, happy to add the utility you suggest. BTW, we don't have good support for complex numbers in JAXopt, see #169.

Reply:
Hmm, I actually just realized that `tree_real` isn't necessary, just `jnp.real` will be enough, since `tree_vdot` returns a scalar.